### DISCUSSION TITLE
RTK Query ServerSideRendering

### DISCUSSION THREAD

**Discussion by barnabasJ** [reduxjs/redux-toolkit#1184] [Ideas]
I'm trying to have serverside rendering together with rtk-query. I read through the [issue]( https://github.com/rtk-incubator/rtk-query/issues/88) on the old repository but I'm not using NextJs nor do I know exactly which components are rendered on the page because this can change depending on a config that I do not control.

I would like to propose the implementation of a similar feature to the [getDataFromTree](https://www.apollographql.com/docs/react/performance/server-side-rendering/#executing-queries-with-getdatafromtree) function from Apollo. In order to do this, it would be necessary to dispatch the action on the server. This does not happen at the moment because `useEffects` aren't exercised on the server. 

It could be possible to add a `ssrMode` flag to `createApi` like the one on the [ApolloClient](https://www.apollographql.com/docs/react/performance/server-side-rendering/#initializing-apollo-client) which leads to the queries being fired once on the server.

I created an example of how I think this could work after the events are also fired once on the server https://github.com/barnabasJ/redux-ssr-example

I also create an example of how it works with apollo here: https://github.com/barnabasJ/apollo-ssr-example

**Comment by markerikson** [comment]
FWIW, anything resembling `getDataFromTree` is almost definitely not something we would do:

- The RTK team has no real SSR experience ourselves
- My understanding is that approach requires some _very_ hacky and unsupported techniques built around hijacking the React rendering process
- Long-term, React's Suspense APIs are going to be the "right" way to handle data fetching in a server environment

I'm open to proposals here, but given our own lack of experience in this area, this is something that would likely require significant involvement by the community - requirements, design, and implementation.

**Reply by barnabasJ** [reply]
I have a working proof of concept. What it is doing is using the renderToStaticMarkup function from react to render the tree. While doing that a middleware listens to the actions in the store. Whenever the middleware encounters a query/pending action it knows it has to wait for the corresponding query/fullfilled or query/rejected action. When all queries were fulfilled or rejected it resolves a promise. Now it is necessary to that again if there could be possible nested components that also run queries.

This definitely has some performance cost, but I think it could be an ok compromise and it is totally optional to use.

If this is something that is potentially ok for you I would clean the code up a bit and open a PR or something where this could be discussed further.

**Reply by phryneas** [reply]
We can certainly take that as a starting point.

Just a counter-question from me. Assuming our hooks would throw a Promise for suspense - would that be supported by SSR/next as well? Because that seems to me like something we'd implement (at least as an option) anyways down the road.

**Reply by barnabasJ** [reply]
I'm not completely sure if these are the exact same use cases as I'm not that familiar with suspense. But as far as I know, Suspense would render the page and e.g. a spinner for components that are still loading in the background, but I'm not sure what happens if you have data that is necessary for the page to load. Would suspense then show a spinner for the whole page? For us, it is more about having the complete page rendered on the server and having it cached somewhere else. But the approach that I took was adding a flag to the `createApi` function which would make the hooks dispatch the action on the server, I think that way it would be possible to support both use cases, having an API slice that collects all required data for rendering and then having other API slices which would load them asynchronously using suspense.

**Reply by phryneas** [reply]
Usually, suspense would prevent rendering of a subtree for a while before showing that spinner, so it is a mechanism that allows a component to signal to the renderer "heyo, I haven't reached my final stage yet". I'm not familiar with SSR implementations, but it seems insane if that mechanism would not be used there.

But yeah, open your PR and I'll try to take a look at it tomorrow and play around with it :)

**Reply by barnabasJ** [reply]
I opened the pr (#1270), please reach out if you have any questions. I'm also on the reactiflux discord (@barnabasj) if you have any questions or if I can help in any way.

**Reply by Ephem** [reply]
This is pretty cool, and what a nice and light proof of concept!

That said, for clarity's sake I'll just be blunt here and say that I think this is the wrong way to go for RTK Query at this point in time for a bunch of reasons I'll try to outline, and I'll try to propose an alternative solution for your specific case as well.

TL;DR: It doesn't play well with where React is heading, and it should be possible to solve in other ways.

**Problems**

1. This solution, and pretty much any other solution for fetching data from within the tree with SSR, will likely break with the new SSR-renderer in React 18.

Leading the community down this path will probably later block them from upgrading to the new renderer (but not React 18 since old renderer will still be around).

2. Suspense is the future, and it's closing in.

> Just a counter-question from me. Assuming our hooks would throw a Promise for suspense - would that be supported by SSR/next as well?

Not out of the box, but this is how you would support SSR data fetching from within the tree in future React. On top of this, if you were to support Suspense, besides the React 18 alpha solutions like [react-ssr-prepass](https://github.com/FormidableLabs/react-ssr-prepass) exist that works today. That uses pretty much the same hacks as the linked PR, but it's third party and does so with Suspense as the basis which is more future proof (and is also a lot faster than `getDataFromTree`, especially in the nested case since it only ever renders one extra time on the server).

> Would suspense then show a spinner for the whole page?

In React pre 18 (excluding experimental builds/release channel), trying to use Suspense on the server directly (without third party solutions like mentioned above) will result in `"ReactDOMServer does not yet support Suspense"` (code [here](https://github.com/facebook/react/blob/v17.0.2/packages/react-dom/src/server/ReactPartialRenderer.js#L936)).

In React 18 the answer depends on a bunch of things, but I don't think we need to go into those here. Suffice to say, even with the new Suspense renderer there will be a way to not stream and instead generate full html just as before, so you will have some control over if and when you want to show spinners.

⚠️ Note that Suspense _for data fetching_ is not currently planned to be included in React 18, so there are no guarantees RTK Query "throwing a Promise" will work fully out of the box, especially with streaming.

3. Hoisting prefetching to the top as the SSR solution also encourages avoiding request waterfalls.

Libraries like RTK Query and React Query has the drawback of nested components having to wait for parents to finish their fetch before they can even start fetching. React Server Components aim to solve this later on, but for now the best practice to avoid request waterfalls on the client is usually to prefetch all queries as early as possible. The current SSR solution encourages this, which is a good thing.

This _is_ harder, but it's also just how React SSR works today if you want to avoid hacks. Maybe in your case it's impossible because of reasons other than technical (you don't own the config), and in other cases maybe it's impractical, but in a strictly technical sense it _should_ always be possible.

4. Because it's a hack, if it can be implemented in an external library, it probably should be.

Since the other issues stem from what the official RTK library wants to endorse and lead the users towards, it would be nice if those that are aware of those tradeoffs and still want to do it could use an external library instead. This also pushes the complexity out of RTK Query.

I think this should be possible, because:

* The `getDataFromTree` function in the PR doesn't rely on any internals.
* While things _are_ placed inside the `useQuerySubscription` hook, at a first glance I think a third party library could work around that part of the functionality by providing their own hooks that dispatch on SSR if a promise doesn't exist, and otherwise delegates to the RTK Query hooks?

I'm not familiar enough with RTK Query to say it's definitely possible though.

**Possible solutions**

_For RTK Query_: Focus on implementing Suspense support instead, even if just experimental. On top of other benefits, with `react-ssr-prepass` this could also solve the usecase @barnabasJ is trying to solve for here. This is also hopefully more future proof, at least for consumers of the library (though heavy refactoring of the library itself might be needed when Suspense for data fetching finally arrives, especially to support streaming ssr).

_For the case where it's hard to hoist prefetching to the top and you are willing to use something hacky/experimental_: `react-ssr-prepass` supports a custom visitor function:

```jsx
ssrPrepass(<App />, (element, instance) => {
  if (element.type === SomeData) {
    return fetchData()
  } else if (instance && instance.fetchData) {
    return instance.fetchData()
  }
})
```

Maybe it would be possible to pair that with prefetching RTK Query-queries in that fetchData @barnabasJ? I haven't tried it myself, but it struck me as a possible way to solve your case.

---

As an aside, I certainly empathize with this problem space and what you are trying to do, so much so that I ended up forking the official SSR-renderer two years ago and implemented Suspense support on top (now long deprecated: [react-lightyear](https://github.com/Ephem/react-lightyear)). 😄

**Reply by Ephem** [reply]
Just to add to point 3. btw, as you mentioned the `getDataFromTree` approach can be _reaally_ slow on the server for the same case (nested fetches) since it renders the app an extra time for every level of nesting.

Many of the users of RTK Query wont know about either the client or the server tradeoff of nested fetches up front and will only see the DX gain of not having to implement prefetching when implementing SSR. When this does become a problem further down the line, it can be hard to refactor out from. This bit a lot of users of Apollo as their applications became more complex as well, something I think we should learn from.

**Reply by markerikson** [reply]
Wow, what a knowledge dump! :)  Thanks, I seriously appreciate you taking the time to write out all this info.  This looks like some extremely valuable to consider here.

**Reply by Ephem** [reply]
My pleasure! Basically what you said with a lot more context (or maybe just words).. 😄

Just to be clear, I definitely think there are valid uses for this approach and it sounds like you have one @barnabasJ. Despite how negative I might sound I think it's a discussion worth having and if it's possible, I think having this as a third party library would be valuable.

This is a really hard and nuanced problem, which is why Server Components took so many years to come up with. 😄

**Reply by Ephem** [reply]
Btw, the way promises are kept track of in this PR looks pretty similar to what is needed to implement Suspense which is another reason I think it makes sense to just go for that all the way instead. Say this approach is implemented first and experimental Suspense comes later, what does that mean for this `getDataFromTree` approach? Would promises be kept track of in multiple places depending on use? Would `getDataFromTree` be deprecated again, or refactored to use Suspense? I can definitely see how this approach could make Suspense more complex to implement in the next step because of having an existing API to support.

**Reply by barnabasJ** [reply]
What a great read, thanks for the feedback. As you have gathered correctly `getDataFromTree` doesn't rely directly on any of the code I added inside `useQuerySubscription`. I only had to add this because I needed a way to dispatch the actions on the server. 

If I understood correctly `getDataFromTree` could also work with the suspense implementation as it only listens to the dispatched actions inside the store and I assume they would still be dispatched with suspense. 
I think it would make total sense to have `getDataFromTree` as a third party library, it just was easier for me to show you what I was trying to accomplish doing this all inside the one repo.

**Reply by phryneas** [reply]
Hmm. Generally, I'd try to keep the "collecting" part out of the hooks, and I think `getDataFromTree` might be the wrong analogy here, since this is very bound to react, which RTKQ itself is not.

I have another angle, hear me out:

Right now, we can't support Suspense because of one icky little timing detail: The promise returned by `dispatch(someEndpoint.invoke())` may resolve with a "this call is currently in progress" action. That makes that promise pretty un-usable for Suspense (and also frustrating to deal with when iteratively invoked in sagas or thunks). 
So I will have to create a central place where all currently running promises will be tracked so I can piggy-back that promise onto the one that's currently running anyways.
I *could* expose those "running queries" via `api.getRunningOperationPromises()` or something.

I think that would solve the `getDataFromTree` question, but more as a side-effect of providing a better user experience.

The other problem would be triggering those `useEffect` calls. And there we could just allow overriding `useEffect` itself maybe? Then a SSR scenario could override `useEffect` with a "just call this" implementation for the hooks.

**Reply by ryota-murakami** [reply]
@phryneas `useEffect()` doesn't run while SSR.

**Reply by phryneas** [reply]
Yes, but "overriding" would also mean "override it with `(cb) => cb()` - meaning a function that just immediately executes

**Reply by phryneas** [reply]
Based on what I've seen in #1270, I have added another PR in #1277 with my suggestion. Do you think that direction could work?

**Reply by Ephem** [reply]
> If I understood correctly getDataFromTree could also work with the suspense implementation as it only listens to the dispatched actions inside the store and I assume they would still be dispatched with suspense.

I think so!

> I think it would make total sense to have getDataFromTree as a third party library, it just was easier for me to show you what I was trying to accomplish doing this all inside the one repo.

Oh yeah, it was a great proof of concept, totally get why you made the PR like that. 😄 If RTK Query would rely on Suspense/throwing promises, one such third party library is `react-ssr-prepass` as I mentioned, but nothing would stop other implementations, or trying out React 18 alphas etc.

> So I will have to create a central place where all currently running promises will be tracked so I can piggy-back that promise onto the one that's currently running anyways.

This definitely sounds like the missing piece of the puzzle!

> The other problem would be triggering those useEffect calls. And there we could just allow overriding useEffect itself maybe? Then a SSR scenario could override useEffect with a "just call this" implementation for the hooks.

I'm not quite sure about this part mostly because I'm not that familiar with the RTK Query implementation, depending on what happens it might not be as easy as just running the same effect, it could mean problems in SSR mode and perhaps a _separate_ "immediateEffect" would be needed. I left a more detailed comment about this with a bunch of questions on the PR (which I think looks good overall!).
